### PR TITLE
handle mutations containing `@defer`

### LINKED ISSUE #2099: `@defer` directive should work with mutations
**Describe the bug**
Mutations should also support `@defer`. Given the following mutation:

```graphql
  mutation ($userId: ID!, $paymentInformation: PaymentInformationInput!) {
    makePayment(userId: $userId, paymentInformation: $paymentInformation) {
      id
      ... @defer {
        paymentStatus {
          __typename
          id
          ... on PaymentSuccess {
            billedAmount
          }
          ... on PaymentFailed {
            reason
          }
        }
      }
    }
  }
```
when following the reproduction steps below, the router returns the error `cannot query field 'makePayment' on type 'Query'`. The same mutation returns a multipart response when used with Apollo Server 4.

I've tried several different syntaxes using `@defer` with mutations in the router but always receive an error.

**To Reproduce**
1. Checkout https://github.com/apollographql/router-defer-e2e-tests/pull/15
2. Follow the README instructions to bring up the router and start the front-end dev server
3. Visit http://localhost:3000/purchase, open the network tab and click the "Make Payment" button
4. Observe the error `cannot query field 'makePayment' on type 'Query'`

**Expected behavior**
A multipart response with the fragment marked with `@defer` arriving after the initial chunk.

### PR DESCRIPTION
**Pull request by Geal**
Fix #2099

to validate the primary part of a request with `@defer`, we reconstruct a partial query by walking through the query plan. The code responsible for that was not detecting mutations.

while the PR in the current state fixes the issue, I'd like to refactor a bit so that we do not special case query and mutation (which will be annoying if we add subscription at some point), and moving the query reconstruction elsewhere, to work directly on the query instead of the query plan

### CODE REVIEW

File: apollo-router/src/services/supergraph_service.rs [hunk 1]
```diff
@@ -710,4 +710,103 @@ mod tests {
...
+            .unwrap()
+            .schema(schema)
+            //.extra_plugin(subgraphs)
+            .build()
+            .await
+            .unwrap();
+
+        let request = supergraph::Request::fake_builder()
+            .header("Accept", "multipart/mixed; deferSpec=20220824")
+            .query(
```
Comment (o0Ignition0o) [inline apollo-router/src/services/supergraph_service.rs:793]: I don't understand, shouldn't we have a userId variable ?

Comment (Geal) [reply apollo-router/src/services/supergraph_service.rs:793]: if it gets to the part that checks the variables, that means it's past the part that was returning the error `cannot query field 'makePayment' on type 'Query'` in #2099, which is what we wanted to fix

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/services/supergraph_service.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router/src/services/supergraph_service.rs:793]: I see, works for me thanks!

**Review by o0Ignition0o — APPROVED** [review]

File: apollo-router/src/services/supergraph_service.rs [hunk 2]
```diff
@@ -710,4 +710,103 @@ mod tests {
 
         insta::assert_json_snapshot!(stream.next_response().await.unwrap());
     }
+
+    #[tokio::test]
+    async fn query_reconstruction() {
+        let schema = r#"schema
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/services/supergraph_service.rs:716]: it's not too large yet, I'd rather have it readable right there with the tests

File: apollo-router/src/services/supergraph_service.rs [hunk 3]
```diff
@@ -710,4 +710,103 @@ mod tests {
...
+  {
+    name: String
+  }
+  "#;
+
+        let service = TestHarness::builder()
+            .configuration_json(serde_json::json!({"include_subgraph_errors": { "all": true } }))
+            .unwrap()
+            .schema(schema)
+            //.extra_plugin(subgraphs)
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/services/supergraph_service.rs]: fixed in eea9a7ad

File: apollo-router/src/query_planner/plan.rs [hunk 4]
```diff
@@ -324,9 +337,16 @@ impl PlanNode {
     }
 }
 
-fn reconstruct_full_query(path: &Path, subselection: &str) -> String {
+fn reconstruct_full_query(path: &Path, kind: &OperationKind, subselection: &str) -> String {
     let mut query = String::new();
     let mut len = 0;
+    let kind_str = match kind {
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/query_planner/plan.rs]: fixed in eea9a7ad

**Pull request merged**